File: IGListTest/IGListTest/Views/WeatherSummaryCell.m
//
//  WeatherSummaryCell.m
//  IGListTest
//
//  Created by liuruixuan on 2023/4/16.
//

#import "WeatherSummaryCell.h"
#import "Utils.h"

@interface WeatherSummaryCell ()



@end

@implementation WeatherSummaryCell

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self.contentView addSubview:self.expandLabel];
        [self.contentView addSubview:self.titleLabel];
        self.contentView.backgroundColor = ssRGBHex(0x0c1f3f);
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    self.titleLabel.frame = CGRectMake(CommonInsets.left, 0, self.titleLabel.bounds.size.width, self.bounds.size.height);
    self.expandLabel.center = CGPointMake(self.bounds.size.width - self.expandLabel.bounds.size.width / 2 - CommonInsets.right, self.bounds.size.height / 2);
    
}

- (void)setExpanded:(BOOL)expanded {
    self.expandLabel.transform = expanded ? CGAffineTransformMakeRotation(M_PI / 2) : CGAffineTransformIdentity;
}

- (UILabel *)expandLabel {
    if (!_expandLabel) {
        _expandLabel = [[UILabel alloc] init];
        _expandLabel.font = AppFont(30);
        _expandLabel.text = @">>";
        _expandLabel.textColor = ssRGBHex(0x44758b);
        _expandLabel.textAlignment = NSTextAlignmentCenter;
        _expandLabel.backgroundColor = [UIColor clearColor];
        [_expandLabel sizeToFit];
    }
    return _expandLabel;
}

- (UILabel *)titleLabel {
    if (!_titleLabel) {
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.backgroundColor = [UIColor clearColor];
        _titleLabel.numberOfLines = 0;
        
        NSMutableParagraphStyle *paragraphStyle = [[NSMutableParagraphStyle alloc] init];
        paragraphStyle.paragraphSpacing = 4;
        NSDictionary *subtitleAttributes = @{
            NSFontAttributeName: AppFont(14),
            NSForegroundColorAttributeName: ssRGBHex(0x42c84b),
            NSParagraphStyleAttributeName: paragraphStyle
        };
        NSDictionary *titleAttributes = @{
            NSFontAttributeName: AppFont(24),
            NSForegroundColorAttributeName: [UIColor whiteColor]
        };
        NSMutableAttributedString *attributedText = [[NSMutableAttributedString alloc] initWithString:@"LATEST\n" attributes:subtitleAttributes];
        [attributedText appendAttributedString:[[NSMutableAttributedString alloc] initWithString:@"WEATHER" attributes:titleAttributes]];
        _titleLabel.attributedText = attributedText;
        [_titleLabel sizeToFit];
    }
    return _titleLabel;
}

@end

File: Marslink_Swift5/Marslink_Swift5/Views/WeatherSummaryCell.swift
//
//  WeatherSummaryCell.swift
//  Marslink_Swift5
//
//  Created by liuruixuan on 2023/4/17.
//

import UIKit

class WeatherSummaryCell: UICollectionViewCell {
    private let expandLabel: UILabel = {
        let label = UILabel()
        label.backgroundColor = UIColor.clear
        label.font = AppFont(size: 30)
        label.textColor = UIColor(hex6: 0x44758B)
        label.textAlignment = .center
        label.text = ">>"
        label.sizeToFit()
        return label
    }()

    let titleLabel: UILabel = {
        let label = UILabel()
        label.backgroundColor = .clear
        label.numberOfLines = 0

        let paragraphStyle = NSMutableParagraphStyle()
        paragraphStyle.paragraphSpacing = 4
        let subtitleAttributes = [
            NSAttributedString.Key.font: AppFont(size: 14),
            NSAttributedString.Key.foregroundColor: UIColor(hex6: 0x42C84B),
            NSAttributedString.Key.paragraphStyle: paragraphStyle,
        ]
        let titleAttributes = [
            NSAttributedString.Key.font: AppFont(size: 24),
            NSAttributedString.Key.foregroundColor: UIColor.white,
        ]
        let attributedText = NSMutableAttributedString(string: "LATEST\n", attributes: subtitleAttributes)
        attributedText.append(NSAttributedString(string: "WEATHER", attributes: titleAttributes))
        label.attributedText = attributedText
        label.sizeToFit()

        return label
    }()

    func setExpanded(_ expanded: Bool) {
        expandLabel.transform = expanded ? CGAffineTransform(rotationAngle: CGFloat.pi / 2) : CGAffineTransform.identity
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        contentView.addSubview(expandLabel)
        contentView.addSubview(titleLabel)
        contentView.backgroundColor = UIColor(hex6: 0x0C1F3F)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        let insets = CommonInsets
        titleLabel.frame = CGRect(x: insets.left, y: 0, width: titleLabel.bounds.width, height: bounds.height)
        expandLabel.center = CGPoint(x: bounds.width - expandLabel.bounds.width / 2 - insets.right, y: bounds.height / 2)
    }
}
